python/cudf/cudf/_libxx/orc.pyx

```diff
@@ -0,0 +1,136 @@
+               timestamp_type=None):
+    """
+    Cython function to call into libcudf API, see `read_orc`.
+
+    See Also
+    --------
+    cudf.io.orc.read_orc
+    """
+    cdef read_orc_args c_read_orc_args = make_read_orc_args(
+        filepath_or_buffer,
+        columns or [],
+        size_t_arg(stripe, "stripe"),
+        size_t_arg(stripe_count, "stripe_count"),
+        size_t_arg(skip_rows, "skip_rows"),
+        size_t_arg(num_rows, "num_rows"),
+        (
+            type_id.EMPTY
+            if timestamp_type is None else
+            np_to_cudf_types[np.dtype(timestamp_type)]
+        ),
+        use_index,
+        decimals_as_float,
+        size_t_arg(force_decimal_scale, "force_decimal_scale")
+    )
+
+    cdef table_with_metadata c_result
+
+    with nogil:
+        c_result = move(read_orc_cpp(c_read_orc_args))
+
+    names = [name.decode() for name in c_result.metadata.column_names]
+
+    return Table.from_unique_ptr(move(c_result.tbl), names)
+
+
+cpdef write_orc(Table table,
+                filepath,
+                compression=None,
+                enable_statistics=False):
+    """
+    Cython function to call into libcudf API, see `write_orc`.
+
+    See Also
+    --------
+    cudf.io.orc.read_orc
+    """
+    cdef compression_type compression_ = compression_type.NONE
+    if compression is None or compression is False:
+        compression_ = compression_type.NONE
+    elif compression == "snappy":
+        compression_ = compression_type.SNAPPY
+    else:
+        raise ValueError(
+            "Unsupported compression type `{}`".format(compression)
+        )
+
+    cdef table_metadata metadata_ = table_metadata()
+    metadata_.column_names.reserve(len(table._column_names))
+
+    for col_name in table._column_names:
+        metadata_.column_names.push_back(str.encode(col_name))
+
+    cdef write_orc_args c_write_orc_args = write_orc_args(
+        sink_info(<string>str(filepath).encode()),
+        table.data_view(), &metadata_, compression_,
+        <bool> (True if enable_statistics else False)
+    )
+
+    with nogil:
+        write_orc_cpp(c_write_orc_args)
+
+
+cdef size_type size_t_arg(arg, name) except*:
+    arg = -1 if arg is None else arg
+    if not isinstance(arg, int) or arg < -1:
+        raise TypeError("{} must be an int >= -1".format(name))
+    return <size_type> arg
+
+
+cdef read_orc_args make_read_orc_args(p, column_names,
+                                      size_type stripe,
+                                      size_type stripe_count,
+                                      size_type skip_rows,
+                                      size_type num_rows,
+                                      type_id timestamp_type,
+                                      bool use_index,
+                                      bool decimals_as_float,
+                                      size_type force_decimal_scale) except*:
+    cdef read_orc_args args = read_orc_args(make_source_info(p))
```

Comment:
Should we be using `make_source_info` instead of `cudf._lib.utils.view_of_buffer` in general now? If so could you update the `json.pyx` and `parquet.pyx` in `_libxx` that are using the prior so that we can phase it out from the ported codebase in general?

Reply:
The choice to use it is up to the implementer, but I factored it out for myself since it seemed like common logic in each of the io bindings. I already suggested this in the Parquet PR https://github.com/rapidsai/cudf/issues/3924, but I can update `json.pyx`.

Reply:
I'll just ask @jdye64 to update both in the Parquet PR, thanks!